### PR TITLE
MONGOID-5587 Auto encryption configuration

### PR DESCRIPTION
**Pull request by comandeo**

### CODE REVIEW

**Review by jamis — CHANGES_REQUESTED** [review]

File: spec/integration/encryption_spec.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
+require 'spec_helper'
+require 'support/crypt/models'
+
+describe 'Encryption' do
+  require_enterprise
```
Comment (johnnyshields) [inline spec/integration/encryption_spec.rb:5]: It appears community edition server supports explicitly-declared encryption (i.e. not auto/schema map style)
https://www.mongodb.com/docs/manual/core/csfle/

File: lib/mongoid/config/encryption.rb [hunk 2]
```diff
@@ -119,15 +126,18 @@ def properties_for_fields(model)
         end
       end
 
-      def properties_for_relations(model)
+      def properties_for_relations(model, visited)
```
[the remark this answers is not in the document]
Comment (comandeo) [reply lib/mongoid/config/encryption.rb:145]: Forgot this one, thank you. Added.

File: lib/mongoid/config/encryption.rb [hunk 3]
```diff
@@ -149,10 +159,10 @@ def algorithm_for(field)
         end
       end
 
-      def key_id_for(key_id_base64)
-        return nil unless key_id_base64
+      def key_id_for(key_id)
```
[the remark this answers is not in the document]
Comment (comandeo) [reply lib/mongoid/config/encryption.rb]: Thank you. Also changed the signature to make it more clear, I hope.

**Review by jamis — APPROVED** [review]

File: spec/integration/encryption_spec.rb (hunk 1 above)
Comment (johnnyshields) [reply spec/integration/encryption_spec.rb:5]: Thank you for your clarification.

One more question: how exactly are schema maps applied? Does the user need to run a migration to apply the latest schema map to the DB? What if two different clients are using two different JSON schema map definitions with accessing the same database?

> What would be your use case where you need to be able to control everything from Rails?

I will need to see if there are any limitations with the auto-encryption approach when it is ready to test. For example, whether it allows me to encrypt existing unencrypted data seamlessly.

**Pull request merged**